src/java_bytecode/select_pointer_type.cpp

```diff
@@ -11,15 +11,101 @@
 /// \param pointer_type: The pointer type replace
 /// \return A pointer type where the subtype may have been modified
 pointer_typet select_pointer_typet::convert_pointer_type(
-  const pointer_typet &pointer_type, const namespacet &ns) const
+  const pointer_typet &pointer_type,
+  const generic_recursion_mapt &generic_recursion_map,
+  const namespacet &ns) const
 {
-  return  pointer_type;
+  // if we have a map of generic parameters -> types and the pointer is
+  // a generic parameter or type, specialize the pointer with concrete types
+  if(
+    !generic_recursion_map.empty() &&
+    (is_java_generic_parameter(pointer_type) ||
+     is_java_generic_type(pointer_type)))
+  {
+    return specialize_generics(pointer_type, generic_recursion_map);
+  }
+  else
+  {
+    return pointer_type;
+  }
+}
+
+/// Find the concrete type assigned to a java generic parameter based on the
+/// current map of parameters -> types.
+/// Example:
+/// - generic parameter: T
+/// - map: T -> U; U -> String
+/// - result: String
+/// \param parameter java generic parameter to be specialized
+/// \param generic_recursion_map map of types for all generic parameters in
+/// the current context
+/// \return the concrete type of the parameter
+pointer_typet select_pointer_typet::find_concrete_type(
+  const pointer_typet &parameter,
+  const generic_recursion_mapt &generic_recursion_map) const
+{
+  PRECONDITION(is_java_generic_parameter(parameter));
+  const irep_idt &parameter_name =
+    to_java_generic_parameter(parameter).get_name();
+
+  if(!generic_recursion_map.count(parameter_name))
+  {
+    // this means that the generic parameter has not been specialized in the
+    // current context; we return the parameter itself which is basically a
+    // pointer to its upper bound
+    return parameter;
+  }
+
+  const pointer_typet &type =
+    generic_recursion_map.find(parameter_name)->second.top();
+
+  // generic parameters can be adopted from outer classes or superclasses so
+  // we may need to search for the concrete type recursively
+  return is_java_generic_parameter(type)
+           ? find_concrete_type(type, generic_recursion_map)
+           : type;
+}
+
+/// Recursively specialize all the generic parameters in the given generic type
+/// based on the current map of parameters -> types.
+/// Example:
+/// - generic type: A<T, B<Integer, U>>
+/// - map: T -> U; U -> String
+/// - result: A<String, B<Integer, String>>
+/// \param type generic type to be specialized
+/// \param generic_recursion_map map of types for all generic parameters in
+/// the current context
+/// \return fully specialized generic type
+pointer_typet select_pointer_typet::specialize_generics(
+  const pointer_typet &type,
+  const generic_recursion_mapt &generic_recursion_map) const
+{
+  if(is_java_generic_parameter(type))
+  {
+    return find_concrete_type(type, generic_recursion_map);
+  }
+  else if(is_java_generic_type(type))
+  {
```

Comment:
I don't follow this block - going to come back to it when I've reviewed the test

Reply:
So if I've understood correctly: this is required because if we have:
```java
class Gen<T> { T field; }
class Gen2<U> { Gen<U> genField; }
Gen2<Integer> param;
```
We need to make sure we map Gen::T -> Gen2::U. `Gen2::U`  is mapped `Integer` in the `generic_recursion_map`. 

What I don't understand is why can't this first relation be in the map also? 

Long term thoughts: when we deal with generic functions, we might have for example that Gen2::U is one of `{Object, Integer, String}`, then here we'd need duplicate however we handle having multiple types. 

Let's discuss this offline tomorrow

Reply:
This is the tree of calls that happen for your example in the code (hope it's not more confusing than helping):
call `gen_nondet_init` on pointer to `Gen2<Integer>` (this adds `Gen2::U -> Integer` to the map)
- call `gen_nondet_pointer_init` on `Gen2<Integer>`
- - call `specialize_generics` on `Gen2<Integer>` (this is java generic type - else if branch)
- - - call `specialize_generics` on `Integer` (this is not parameter, not generic type - else branch)
- - call `gen_nondet_init` on class `Gen2`
- - - call `gen_nondet_struct_init` on `Gen2`
- - - - call `gen_nondet_init` on pointer to `Gen<U>` as a field of `Gen2` (this adds `Gen::T -> Gen2::U` to the map)
- - - - - call `gen_nondet_pointer_init` on `Gen<U>`
- - - - - - call `specialize_generics` on `Gen<U>` (this is java generic type - else if branch)
- - - - - - - call `specialize_generics` on `U` (this is parameter - if branch) which returns `Integer`
- - - - - - call `gen_nondet_init` on class `Gen`
- - - - - - - call `gen_nondet_struct_init` on `Gen`
- - - - - - - - call `gen_nondet_init` on pointer to `T` as field of `Gen`
- - - - - - - - - call `gen_nondet_pointer_init` on `T`
- - - - - - - - - - call `specialize_generics` on `T` (this is parameter - if branch) which returns `Integer`

So you do have all you need in the map. In `specialize_generics`, you do not manipulate the map, just recursively go over the generic type (might have multiple and/or nested generic types and parameters in it such as `Gen<GenTwo<U, Integer>>`).

Reply:
That is helpful, but I don't see how the result is used anywhere? Seems like the first call adds `Get2::U -> Integer` to the map, so we don't care when we read it on 

>               - call specialize_generics on U (this is parameter - if branch) which returns Integer

Reply:
After an offline discussion, I believe this indeed can be reduced to only specialize generics if the type at hand is a generic parameter.